anndata/utils.py

```diff
@@ -59,6 +59,54 @@ def convert_to_dict_nonetype(obj: None):
     return dict()
 
 
+@singledispatch
+def dim_len(x, dim):
+    """\
+    Return the size of an array in dimension `dim`.
+
+    Returns None if `x` is an awkward array with variable length in the requested dimension.
+    """
+    return x.shape[dim]
+
+
+try:
+    from .compat import awkward as ak
+
+    @dim_len.register(ak.Array)
+    def dim_len_awkward(x, dim):
+        if dim == 0:
+            # dimension 0 is a special case - it is always of `ArrayType` and has a fixed length.
+            try:
+                return x.type.length
+            except AttributeError:
+                raise ValueError("The outermost type must be an `awkward.Array`!")
```

Comment:
If we're dispatching on `ak.Array` how could the outermost type not be an `Array`?